pandas/core/missing.py

```diff
@@ -964,3 +974,107 @@ def _rolling_window(a: np.ndarray, window: int):
     shape = a.shape[:-1] + (a.shape[-1] - window + 1, window)
     strides = a.strides + (a.strides[-1],)
     return np.lib.stride_tricks.as_strided(a, shape=shape, strides=strides)
+
+
+def _can_hold_element(values, element: Any) -> bool:
+    """
+    Expanded version of core.dtypes.cast.can_hold_element
+    """
+    from pandas.core.arrays import (
+        ExtensionArray,
+        FloatingArray,
+        IntegerArray,
+    )
+
+    if isinstance(values, ExtensionArray):
+        if hasattr(values, "_validate_setitem_value"):
+            # NDArrayBackedExtensionArray
+            try:
+                # error: "Callable[..., Any]" has no attribute "_validate_setitem_value"
+                values._validate_setitem_value(element)  # type: ignore[attr-defined]
+                return True
+            except (ValueError, TypeError):
+                return False
+        else:
+            # other ExtensionArrays
+            return True
+    else:
+        element = extract_array(element, extract_numpy=True)
+        if isinstance(element, (IntegerArray, FloatingArray)):
+            if element._mask.any():
+                return False
+        return can_hold_element(values, element)
+
+
+def coerce_to_target_dtype(values, other):
+    """
+    Coerce the values to a dtype compat for other. This will always
+    return values, possibly object dtype, and not raise.
+    """
+    dtype, _ = infer_dtype_from(other, pandas_dtype=True)
+    new_dtype = find_common_type([values.dtype, dtype])
+
+    return astype_array_safe(values, new_dtype, copy=False)
+
+
+def fillna_array(values, value, limit=None, inplace: bool = False, downcast=None):
+    from pandas.core.array_algos.putmask import (
+        putmask_inplace,
+        validate_putmask,
+    )
+    from pandas.core.arrays import ExtensionArray
+
+    # inplace = validate_bool_kwarg(inplace, "inplace")
+
+    if isinstance(values, ExtensionArray):
+        if (
+            not _can_hold_element(values, value)
+            and values.dtype.kind != "m"
+            and not is_categorical_dtype(values.dtype)
+        ):
+            # We support filling a DatetimeTZ with a `value` whose timezone
+            #  is different by coercing to object.
+            # TODO: don't special-case td64
+            values = values.astype(object)
+            return fillna_array(
+                values, value, limit=limit, inplace=True, downcast=downcast
+            )
+
+        values = values if inplace else values.copy()
```

Comment:
i think EA.fillna is never inplace, so this may be making an unnecessary copy (or not being inplace even when the user specifies it) (applicable to status quo too)